Add FirestoreStore backend

alexeylark opened the pull request:
### Motivation
We needed a Firestore implementation for our FastMCP build on GCP, and developed FirestoreStore as an extension of `py-key-value`, which has been working for our case. Now we're contributing the code, so it's available from the main repository for everybody to use. 

### Summary

 Adds a Firestore-backed store implementation (FirestoreStore) to py-key-value-aio, with corresponding tests, packaging extras, docs updates, and generated sync code via make codegen.

  ### Installation

  pip install py-key-value-aio[firestore]

  ### Usage

 
```
python from key_value.aio.stores.firestore import FirestoreStore

store = FirestoreStore(credentials=google_credentials, database="firestore-db")
```


  ### Tests / Validation

  - make precommit
  - make test-aio-concise
  - Built and installed a custom wheel and validated integration in FastMCP==2.14 using FernetEncryptionWrapper (end-to-end smoke test).

  ### Notes

  - Sync library changes are generated from aio (make codegen); sync files are included in this PR as expected.
  - Some formatting tests for documentation may show errors. 
  - Happy to adjust behavior or naming to match project conventions — please request changes if you’d like it shaped differently.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added Firestore as a supported backend for the key-value store, available in both async and sync implementations.

* **Documentation**
  * Updated installation guides to include Firestore setup with new pip install extras.
  * Added comprehensive Firestore usage documentation and backend support details.

* **Tests**
  * Added emulator-based test suite for Firestore integration.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

alexeylark commented on key-value/key-value-aio/src/key_value/aio/stores/firestore/store.py:
```diff
@@ -0,0 +1,122 @@
+    async def _close(self) -> None:
+        """Close the Firestore client."""
+        if self._client and not self._client_provided_by_user:
+            self._client.close()
```
Google Client is closed synchronously. This is method from the source code:
```python
class Client(_ClientFactoryMixin):    
    def close(self):
        """Clean up transport, if set.

        Suggested use:

        .. code-block:: python

           import contextlib

           with contextlib.closing(client):  # closes on exit
               do_something_with(client)
        """
        if self._http_internal is not None:
            self._http_internal.close()
```

strawgate commented:
Thanks, will try to find time to review this week. I've enabled the tests which may require some follow-up before this is ready

alexeylark commented:
@strawgate 
Hi!
I added a couple of commits:
- fixed the excessive `@override`.
- blocked the Pyright errors warning about untyped output coming from third-party library (google-cloud-firestore).
- changed the google-credentials object annotation.
- fixed the table indentation in `.md` files.

I reviewed the output from GitHub tests and from CodeRabbit. 
- GitHub errors are not linked to Firestore implementation.
- Firestore implementation at the moment raises one error suggestion by CodeRabbit, mentioning Firestore's `AsyncClient` closed synchronously - I checked, and the client is actually closed synchronously in this module, so I think the suggestion by the agent is wrong. 
- ran `make precommit` on the dev-container showing no errors. 

Please let me know if you spot anything else that needs changing.

strawgate commented on docs/wrappers.md:
```diff
@@ -7,7 +7,7 @@ protocol, so they can be used anywhere a store can be used.
 
 | Wrapper | Description |
-|---------|-------------|
+| ------- | ----------- |
```
remove unrelated changes

strawgate commented on key-value/key-value-aio/src/key_value/aio/stores/firestore/store.py:
```diff
@@ -0,0 +1,121 @@
+    async def _get_managed_entry(self, *, key: str, collection: str | None = None) -> ManagedEntry | None:
+        """Get a managed entry from Firestore."""
+        collection = collection or self.default_collection
+        response = await self._connected_client.collection(collection).document(key).get()  # pyright: ignore[reportUnknownMemberType]
```
Is there a type hint for this?

alexeylark replied:
There was a typehint `DocumentSnapshot`, but it throws `Type of "get" is partially unknown` aggressively, and I blame the Google library there, because the typehints go to the `Any` in the baseclasses. I may be wrong here, but that's what I ended up with.

strawgate commented on key-value/key-value-aio/tests/stores/firestore/test_firestore.py:
```diff
@@ -0,0 +1,132 @@
+        return _InMemoryAsyncFirestoreDocument(storage=self._storage, collection=self._name, key=key)
+
+
+class InMemoryAsyncFirestoreClient(firestore.AsyncClient):
```
We dont mock clients for tests, a mock client doesnt actually test anything. We'd want to use something like https://firebase.google.com/docs/emulator-suite to spin up a docker instance that mimics a real firebase deployment.

strawgate commented on key-value/key-value-aio/src/key_value/aio/stores/firestore/store.py:
```diff
@@ -0,0 +1,121 @@
+    from google.cloud import firestore
+    from google.oauth2.service_account import Credentials
+except ImportError as e:
+    msg = "FirestoreStore requires py-key-value-aio[firestore]"
```
Due to a quirk in how codegen runs let's just drop the reference to the package name

```suggestion
    msg = "FirestoreStore requires the `firestore` extra"
```

strawgate reviewed (dismissed):
We dont currently accept stores where we can't validate against a live/test system. We can try the firebase emulator or I can maybe see about signing up for a firebase account and using a key for tests. Without an emulator/docker firebase system we just have no way to ensure we aren't breaking things in future updates.

alexeylark commented:
> @alexeylark i'd be happy to take a stab at this for a release late next week, but it would probably require me to take over the PR. Let me know if you have time to take a look.

@strawgate Sure, I'll go over it this week, thank you!

alexeylark commented on docs/stores.md:
```diff
@@ -30,7 +30,7 @@ long-term storage, prefer stable stores.
 
 | Store | Stability | Async | Sync | Description |
-|-------|:---------:|:-----:|:----:|:------------|
+| ----- | :-------: | :---: | :--: | ----------- |
```
Removed.

strawgate commented:
@macroscopeapp can you review this PR?

strawgate commented:
Claude was wrong about async close, but right about using dict over string if we can. 

I've updated the PR to not stringify the contents when writing to Firestore. Please review my latest commit and If this is good with you we can merge

strawgate commented on key-value/key-value-aio/src/key_value/aio/stores/firestore/store.py:
```diff
@@ -0,0 +1,114 @@
+        """Delete a managed entry from Firestore."""
+        collection = collection or self.default_collection
+        await self._client.collection(collection).document(key).delete()  # pyright: ignore[reportUnknownMemberType,reportOptionalMemberAccess]
+        return True
```
Fix it for me, along with the other feedback. The pattern we follow is we check if it exists to determine if we should return true or false but we call the delete anyway.

strawgate commented:
im doing some refactoring on main so don't worry about fixing conflicts, ill take care of them

paulo-raca commented:
Thank you!
